cudamapper/src/chainer_utils.cu

```diff
@@ -0,0 +1,221 @@
+                                             int32_t num_overlaps)
+{
+    int32_t d_thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+    int32_t stride      = blockDim.x * gridDim.x;
+    for (int i = d_thread_id; i < num_overlaps; i += stride)
+    {
+        int32_t chain_start  = chain_starts[i];
+        int32_t chain_length = chain_lengths[i];
+        for (int32_t ind = chain_start; ind < chain_start + chain_length; ++i)
+        {
+            anchor_chains[ind] = ind;
+        }
+    }
+}
+
+__global__ void output_overlap_chains_by_backtrace(const Overlap* overlaps,
+                                                   const Anchor* anchors,
+                                                   const bool* select_mask,
+                                                   const int32_t* predecessors,
+                                                   int32_t* anchor_chains,
+                                                   int32_t* anchor_chain_starts,
+                                                   int32_t num_overlaps,
+                                                   bool check_mask)
+{
+    int32_t d_thread_id = blockIdx.x * blockDim.x + threadIdx.x;
+    int32_t stride      = blockDim.x * gridDim.x;
+
+    // Processes one overlap per iteration,
+    // "i" corresponds to an overlap
+    for (int i = d_thread_id; i < num_overlaps; i += stride)
+    {
+        // index within this chain of anchors (i.e., the anchors within a single overlap)
+
+        if (!check_mask || (check_mask & select_mask[i]))
+        {
+            int32_t anchor_chain_index = 0;
+            // As chaining proceeds backwards (i.e., it's a backtrace),
+            // we need to fill the new anchor chain array in in reverse order.
+            int32_t index = anchor_chain_starts[i];
+            while (index != -1)
+            {
+                anchor_chains[anchor_chain_starts[i] + (overlaps[i].num_residues_ - anchor_chain_index)] = index;
+                int32_t pred                                                                             = predecessors[index];
+                index                                                                                    = pred;
+                ++anchor_chain_index;
+            }
+        }
+    }
+}
+
+__global__ void backtrace_anchors_to_overlaps(const Anchor* anchors,
+                                              Overlap* overlaps,
+                                              double* scores,
+                                              bool* max_select_mask,
+                                              int32_t* predecessors,
+                                              const int32_t n_anchors,
+                                              const int32_t min_score)
+{
+    const std::size_t d_tid = blockIdx.x * blockDim.x + threadIdx.x;
+    if (d_tid < n_anchors)
+    {
+
+        int32_t global_overlap_index = d_tid;
```

Comment:
`const` (and in many other places)

Reply:
Resolved by 78da54195e1d117af4d7399b7fd222304a6b9316.